becquerel/core/peakfinder.py

```diff
@@ -0,0 +1,395 @@
+    and negative wings to subtract the continuum, e.g., a boxcar:
+
+    +2|     ┌───┐     |
+      |     │   │     |
+     0|─┬───┼───┼───┬─|
+    -1| └───┘   └───┘ |
+
+    The positive part is meant to detect the peak, and the negative part to
+    sample the continuum under the peak.
+
+    The kernel should sum to 0.
+
+    The width of the kernel scales proportionally to the square root
+    of the channels, with a minimum value set by 'fwhm_at_0'.
+
+    """
+
+    def __init__(self, ref_channel, ref_fwhm, fwhm_at_0=1.):
+        """Initialize with a reference line position and FWHM in channels."""
+        if ref_channel <= 0:
+            raise PeakFilterError('Reference channel must be positive')
+        if ref_fwhm <= 0:
+            raise PeakFilterError('Reference FWHM must be positive')
+        if fwhm_at_0 < 0:
+            raise PeakFilterError('FWHM at 0 must be non-negative')
+        self.ref_channel = float(ref_channel)
+        self.ref_fwhm = float(ref_fwhm)
+        self.fwhm_at_0 = float(fwhm_at_0)
+
+    def fwhm(self, channel):
+        """Calculate the expected FWHM at the given channel."""
+        # f(x)^2 = f0^2 + k x^2
+        # f1^2 = f0^2 + k x1^2
+        # k = (f1^2 - f0^2) / x1^2
+        # f(x)^2 = f0^2 + (f1^2 - f0^2) (x/x1)^2
+        f0 = self.fwhm_at_0
+        f1 = self.ref_fwhm
+        x1 = self.ref_channel
+        fwhm_sqr = f0**2 + (f1**2 - f0**2) * (channel / x1)**2
+        return np.sqrt(fwhm_sqr)
+
+    def kernel(self, channel, n_channels):
+        """Generate the kernel for the given channel."""
+        raise NotImplementedError
+
+    def kernel_matrix(self, n_channels):
+        """Build a matrix of the kernel evaluated at each channel."""
+        kern = np.zeros((n_channels, n_channels))
+        for j in range(n_channels):
+            kern[:, j] = self.kernel(j, n_channels)
+        kern_pos = +1 * kern.clip(0, np.inf)
+        kern_neg = -1 * kern.clip(-np.inf, 0)
+        # normalize negative part to be equal to the positive part
+        kern_neg *= kern_pos.sum(axis=0) / kern_neg.sum(axis=0)
+        return kern_pos - kern_neg
+
+    def plot_matrix(self, n_channels):
+        """Plot the matrix of kernels evaluated across the channels."""
+        kern_mat = self.kernel_matrix(n_channels)
+        kern_min = kern_mat.min()
+        kern_max = kern_mat.max()
+        kern_min = min(kern_min, -1 * kern_max)
+        kern_max = max(kern_max, -1 * kern_min)
+
+        plt.imshow(
+            kern_mat.T[::-1, :], cmap=plt.get_cmap('bwr'),
+            vmin=kern_min, vmax=kern_max,
+            extent=[n_channels, 0, 0, n_channels])
+        plt.colorbar()
+        plt.xlabel('Input channel')
+        plt.ylabel('Output channel')
+        plt.gca().set_aspect('equal')
+
+    def convolve(self, data):
+        """Convolve this kernel with the data."""
+        kern_mat = self.kernel_matrix(len(data))
+        kern_mat_pos = +1 * kern_mat.clip(0, np.inf)
```

Comment:
do we want to do, e.g. kern_mat.clip(0, np.inf) # * (+1), so we don't have to do an extra unnecessary multiplication operation?

Reply:
The multiplication by +1 is for parallelism with the line below (see also line 71).